Content.Server/Power/EntitySystems/BatterySystem.cs

```diff
@@ -93,7 +111,53 @@ private void CalculateBatteryPrice(EntityUid uid, BatteryComponent component, re
         private void OnEmpPulse(EntityUid uid, BatteryComponent component, ref EmpPulseEvent args)
         {
             args.Affected = true;
-            component.UseCharge(args.EnergyConsumption);   
+            UseCharge(uid, args.EnergyConsumption, component);
+        }
+
+        public float UseCharge(EntityUid uid, float value, BatteryComponent? battery = null)
+        {
+            if (!Resolve(uid, ref battery))
+                return 0;
+
+            var newValue = Math.Clamp(0, battery.CurrentCharge - value, battery._maxCharge);
+            var delta = newValue - battery.Charge;
+            battery.Charge = newValue;
+            var ev = new ChargeChangedEvent(battery.CurrentCharge, battery._maxCharge);
+            RaiseLocalEvent(uid, ref ev);
+            return delta;
+        }
```

Comment:
If ChargeChangedEvent should only be raised if it actually changed, you need an early return for when `delta` is zero.

Same thing for SetCharge/SetMaxCharge.